pkg/memo/memo.go

```diff
@@ -0,0 +1,73 @@
+package memo
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+// InboundMemo represents the inbound memo structure for non-EVM chains
+type InboundMemo struct {
+	// Header contains the memo header
+	Header
+
+	// FieldsV0 contains the memo fields V0
+	// Note: add a FieldsV1 if breaking change is needed in the future
+	FieldsV0
+}
+
+// EncodeToBytes encodes a InboundMemo struct to raw bytes
+//
+// Note:
+//   - Any provided 'DataFlags' is ignored as they are calculated based on the fields set in the memo.
+//   - The 'RevertGasLimit' is not used for now for non-EVM chains.
+func (m *InboundMemo) EncodeToBytes() ([]byte, error) {
+	// encode head
+	head, err := m.Header.EncodeToBytes()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to encode memo header")
+	}
+
+	// encode fields based on version
+	var data []byte
+	switch m.Version {
+	case 0:
+		m.DataFlags, data, err = m.FieldsV0.Pack(m.OpCode, m.EncodingFormat)
+	default:
+		return nil, fmt.Errorf("invalid memo version: %d", m.Version)
+	}
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to pack memo fields version: %d", m.Version)
+	}
+
+	// update data flags with the calculated value
+	head[3] = m.DataFlags
```

Comment:
Consider moving `3` to `const headerIndex...`

Reply:
Yeah. I updated this logic by explicit assignment: `m.Header.DataFlags = dataFlags`